pkg/util/pki/csr_test.go

```diff
@@ -1079,3 +1084,109 @@ func TestEncodeX509Chain(t *testing.T) {
 		})
 	}
 }
+
+func rsaKey(t *testing.T, size int) crypto.Signer {
+	t.Helper()
+
+	key, err := rsa.GenerateKey(rand.Reader, size)
+	if err != nil {
+		t.Fatalf("failed to generate RSA key with size %d: %s", size, err)
+	}
+
+	return key
+}
+
+func ecdsaKey(t *testing.T, curve elliptic.Curve) crypto.Signer {
+	t.Helper()
+
+	key, err := ecdsa.GenerateKey(curve, rand.Reader)
+	if err != nil {
+		t.Fatalf("failed to generate ECDSA key with curve %s: %s", curve, err)
+	}
+
+	return key
+}
+
+func ed25519Key(t *testing.T) crypto.Signer {
+	t.Helper()
+
+	_, priv, err := ed25519.GenerateKey(rand.Reader)
+	if err != nil {
+		t.Fatalf("failed to generate ed25519 key: %s", err)
+	}
+
+	return priv
+}
+
+func Test_SignCertificate_Signatures(t *testing.T) {
+	specs := map[string]struct {
+		SignerKey                  crypto.Signer
+		ExpectedSignatureAlgorithm x509.SignatureAlgorithm
+	}{
+		"RSA 2048": {
+			SignerKey:                  rsaKey(t, 2048),
+			ExpectedSignatureAlgorithm: x509.SHA256WithRSA,
+		},
+		"RSA 3072": {
+			SignerKey:                  rsaKey(t, 3072),
+			ExpectedSignatureAlgorithm: x509.SHA384WithRSA,
+		},
+		"RSA 4096": {
+			SignerKey:                  rsaKey(t, 4096),
+			ExpectedSignatureAlgorithm: x509.SHA512WithRSA,
+		},
```

Comment:
Could we add RSA > 4096 as a testcase?

Reply:
We can, although they only get slower to generate. I suppose adding one won't hurt.

Something I thought while doing this was that it would be nice to have a test helper which uses `sync.Once` or similar to generate a single key that can be shared across multiple tests which need _a_ key but don't need anything specific. Obviously not for this PR though.